File: files/CH-113.R
library(tidyverse)
library(readxl)

path = "files/CH-113 Manage Duplicate Values.xlsx"
input = read_excel(path, range = "B2:B15")
test  = read_excel(path, range = "D2:D15")

result <-  input %>%
  mutate(rn = row_number()) %>%
  arrange(`Product ID`) %>%
  mutate(dup = n() > 1,
         a = cumsum(rn - lag(rn, default = first(rn)) != 1),
         .by = `Product ID`) %>%
  mutate(b = row_number(), .by = c(`Product ID`, a)) %>%
  arrange(rn) %>%
  mutate(result = ifelse(dup, paste0(`Product ID`,"-",a, "-", b), `Product ID`)) 


identical(result$result, test$`Product ID`)
#> [1] TRUE

File: files/100-199/113/CH-113.py
import pandas as pd

path = "CH-113 Manage Duplicate Values.xlsx"

input = pd.read_excel(path, usecols="B", skiprows=1, names=["Product ID"])
test = pd.read_excel(path, usecols="D", skiprows=1, names=["result"])

input['rn'] = input.index + 1
input = input.sort_values('Product ID').reset_index()
input['dup'] = input.groupby('Product ID')['Product ID'].transform('size') > 1
input['a'] = input.groupby('Product ID')['rn'].diff().gt(1).cumsum()
input['a'] = input.groupby('Product ID')['a'].transform(lambda x: x - x.min() + 1)
input['b'] = input.groupby(['Product ID', 'a']).cumcount() + 1
input = input.sort_values('rn').reset_index()
input['result'] = input.apply(lambda row: f"{row['Product ID']}-{row['a']}-{row['b']}" if row['dup'] else row['Product ID'], axis=1)
print(input['result'].equals(test["result"])) # True
